src/lib/paginator/paginator.ts

```diff
@@ -76,26 +76,42 @@ export class MatPaginator extends _MatPaginatorBase implements OnInit, OnDestroy
   @Input()
   get pageIndex(): number { return this._pageIndex; }
   set pageIndex(value: number) {
-    this._pageIndex = Math.max(coerceNumberProperty(value), 0);
-    this._changeDetectorRef.markForCheck();
+    const newPageIndex = Math.max(coerceNumberProperty(value), 0);
+    const oldPageIndex = this._pageIndex;
+
+    if (newPageIndex !== oldPageIndex) {
+      this._pageIndex = newPageIndex;
+      this._emitPageEvent(oldPageIndex);
+      this._changeDetectorRef.markForCheck();
+    }
   }
   _pageIndex: number = 0;
 
   /** The length of the total number of items that are being paginated. Defaulted to 0. */
   @Input()
   get length(): number { return this._length; }
   set length(value: number) {
-    this._length = coerceNumberProperty(value);
-    this._changeDetectorRef.markForCheck();
+    const newLength = coerceNumberProperty(value);
+
+    if (newLength !== this._length) {
+      this._length = newLength;
+      this._emitPageEvent(this.pageIndex);
```

Comment:
Since alot of usages of the `MatPaginator` with a `DataSource` update the length whenever the data updates, this could cause a loop of sorts.

The if block would prevent it from happening more than once, but we still would be triggering the `DataSource`'s value multiple times.

An example can be seen [here](https://github.com/angular/material2/blob/master/src/lib/table/table-data-source.ts#L270) with the `table-data-source`.

Reply:
Doesn't the table user the `IterableDiffer` under the hood? That should prevent it from doing extra work even if it does fire multiple times. I'm fine with cutting down the PR so it only addresses the `pageSize` as mentioned in #12583, but I can imagine people running into the same issue for the other two properties.

Reply:
We just got another issue about the `pageIndex`: https://github.com/angular/material2/issues/12620.